instrumentation/armeria/armeria-1.3/library/src/main/java/io/opentelemetry/instrumentation/armeria/v1_3/ArmeriaTelemetryBuilder.java

```diff
@@ -11,87 +11,48 @@
+  private final DefaultHttpClientInstrumenterBuilder<ClientRequestContext, RequestLog>
+      clientBuilder;
+  private final DefaultHttpServerInstrumenterBuilder<ServiceRequestContext, RequestLog>
+      serverBuilder;
 
-  private final OpenTelemetry openTelemetry;
-  @Nullable private String peerService;
-  private boolean emitExperimentalHttpClientMetrics = false;
-  private boolean emitExperimentalHttpServerMetrics = false;
-
-  private final List<AttributesExtractor<? super RequestContext, ? super RequestLog>>
-      additionalExtractors = new ArrayList<>();
-  private final List<AttributesExtractor<? super ClientRequestContext, ? super RequestLog>>
-      additionalClientExtractors = new ArrayList<>();
-
-  private final HttpClientAttributesExtractorBuilder<RequestContext, RequestLog>
-      httpClientAttributesExtractorBuilder =
-          HttpClientAttributesExtractor.builder(ArmeriaHttpClientAttributesGetter.INSTANCE);
-  private final HttpServerAttributesExtractorBuilder<RequestContext, RequestLog>
-      httpServerAttributesExtractorBuilder =
-          HttpServerAttributesExtractor.builder(ArmeriaHttpServerAttributesGetter.INSTANCE);
-
-  private final HttpSpanNameExtractorBuilder<RequestContext> httpClientSpanNameExtractorBuilder =
-      HttpSpanNameExtractor.builder(ArmeriaHttpClientAttributesGetter.INSTANCE);
-  private final HttpSpanNameExtractorBuilder<RequestContext> httpServerSpanNameExtractorBuilder =
-      HttpSpanNameExtractor.builder(ArmeriaHttpServerAttributesGetter.INSTANCE);
-  private Function<
-          SpanNameExtractor<RequestContext>, ? extends SpanNameExtractor<? super RequestContext>>
-      clientSpanNameExtractorTransformer = Function.identity();
-  private Function<
-          SpanNameExtractor<RequestContext>, ? extends SpanNameExtractor<? super RequestContext>>
-      serverSpanNameExtractorTransformer = Function.identity();
-
-  private final HttpServerRouteBuilder<RequestContext> httpServerRouteBuilder =
-      HttpServerRoute.builder(ArmeriaHttpServerAttributesGetter.INSTANCE);
-
-  private Function<
-          SpanStatusExtractor<RequestContext, RequestLog>,
-          ? extends SpanStatusExtractor<? super RequestContext, ? super RequestLog>>
-      statusExtractorTransformer = Function.identity();
+  static {
+    ArmeriaInstrumenterBuilderUtil.setClientBuilderExtractor(
+        ArmeriaTelemetryBuilder::getClientBuilder);
+    ArmeriaInstrumenterBuilderUtil.setServerBuilderExtractor(
+        ArmeriaTelemetryBuilder::getServerBuilder);
+  }
 
+  @SuppressWarnings({"rawtypes", "unchecked"})
   ArmeriaTelemetryBuilder(OpenTelemetry openTelemetry) {
-    this.openTelemetry = openTelemetry;
+    clientBuilder = ArmeriaInstrumenterBuilderFactory.getClientBuilder(openTelemetry);
+    serverBuilder = ArmeriaInstrumenterBuilderFactory.getServerBuilder(openTelemetry);
   }
 
+  @SuppressWarnings({"unchecked", "rawtypes"})
```

Comment:
I think this is pointing to an issue, I don't think the same status extractor function can apply to both `ClientRequestContext` and `ServiceRequestContext`

Reply:
they extend the same superclass.

I just can't get the correct generic signature - and ideally that is not a breaking change

Reply:
oh, I see, thanks

> I just can't get the correct generic signature

I tried too 😭